File: lib/io.cc
/*
 * inci driver
 * Copyright (C) 2008 i Bilisim Teknolojileri Ltd. Sti.
 *
 */

#include <unistd.h>
#include <stdlib.h>

#include "io.h"

static unsigned char
ibutton_crc(const unsigned char *buf, size_t len)
{
	unsigned char crc = 0;

	for (unsigned int pos = 0; pos < len; ++pos) {
		crc = crc ^ buf[pos];
		for (int i = 0; i < 8; i++) {
			if (crc & 0x01)
				crc = (crc >> 1) ^ 0x8C;
			else
				crc >>= 1;
		}
	}
	return crc;
}


iO::iO()
{
	inpos = 0;

	// Opening serial port activates iO
	port = new SeriPort("/dev/ttyS0", 38400);
	if (!port->is_ok) { errno = -1; return; }

	// Give iO some time to setup UART
	usleep(500000); // half a second

	handshake();
	if (errno) return;

	errno = 0;
}

iO::~iO()
{
	delete port;
}

int
iO::reconnect(void)
{
	const char override[] = "\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00";

	// Send max command size of zeroes, this puts firmware back
	// to the command state.
	port->write(override, sizeof(override) - 1);
	// Flush the input buffer, it may contain previous garbage
	inpos = 0;
	port->flush_input();
	// Handshake again
	handshake();

	return errno == 0;
}

void
iO::set_dio(unsigned int mask, unsigned int state)
{
	outbuf[2] = mask & 0xff;
	outbuf[3] = (mask & 0xff00) >> 8;
	outbuf[4] = (mask & 0xff0000) >> 16;
	outbuf[5] = (mask & 0xff000000) >> 24;
	outbuf[6] = state & 0xff;
	outbuf[7] = (state & 0xff00) >> 8;
	outbuf[8] = (state & 0xff0000) >> 16;
	outbuf[9] = (state & 0xff000000) >> 24;

	send('o', 8);
}

void
iO::set_servo(int servo)
{
	outbuf[2] = 0x01;
	outbuf[3] = servo & 0xff;
	outbuf[4] = (servo & 0xff00) >> 8;
	send('s', 3);
}

void
iO::set_motor_pwm(int fl, int fr, int rear)
{
	unsigned int mask;
	unsigned int state = 0;

	mask = (1 << 13) | (1 << 14) | (1 << 21) | (1 << 22) | (1 << 23) | (1 << 24);

	if (fl >= 0) state |= (1 << 14); else state |= (1 << 13);
	if (fr >= 0) state |= (1 << 22); else state |= (1 << 21);
	if (rear >= 0) state |= (1 << 24); else state |= (1 << 23);
	set_dio(mask, state);

	outbuf[2] = 0x01 | 0x02 | 0x04;
	outbuf[3] = abs(fl);
	outbuf[4] = abs(fr);
	outbuf[5] = abs(rear);
	send('p', 4);
}

void
iO::read_report(struct iO_data &data)
{
	send('r', 0);
inpos = 0;
	while (0 == recv())
		;
inpos = 0;
	if (inbuf[1] != 'd') return;
std::cout << "got it" << std::endl;
	data.encoder[0] = (inbuf[2] << 8) + inbuf[3];
	data.encoder[1] = (inbuf[4] << 8) + inbuf[5];
	data.encoder[2] = (inbuf[6] << 8) + inbuf[7];

	data.adc[0] = inbuf[8 + 0] + ((inbuf[8 + 4] & 0xc0) << 2);
	data.adc[1] = inbuf[8 + 1] + ((inbuf[8 + 4] & 0x30) << 4);
	data.adc[2] = inbuf[8 + 2] + ((inbuf[8 + 4] & 0x0c) << 6);
	data.adc[3] = inbuf[8 + 3] + ((inbuf[8 + 4] & 0x03) << 8);
	data.adc[4] = inbuf[8 + 5] + ((inbuf[8 + 9] & 0xc0) << 2);
	data.adc[5] = inbuf[8 + 6] + ((inbuf[8 + 9] & 0x30) << 4);
	data.adc[6] = inbuf[8 + 7] + ((inbuf[8 + 9] & 0x0c) << 6);
	data.adc[7] = inbuf[8 + 8] + ((inbuf[8 + 9] & 0x03) << 8);
	data.adc[8] = inbuf[8 + 10] + (inbuf[8 + 11] << 8);

	data.dio = inbuf[20]
		+ (inbuf[21] << 8)
		+ (inbuf[22] << 16)
		+ (inbuf[23] << 24);
}

void
iO::handshake(void)
{
	send('v', 0);
	while (0 == recv())
		;
	if (inbuf[1] != 'i' || inbuf[2] != 'O') {
		errno = -5;
		return;
	}
	version = inbuf[3];
	revision = inbuf[4];
}

int
iO::recv()
{
	int ret;
	int need;

	if (inpos < 2) need = 2 - inpos;
	else {
		if (inbuf[0] != 'R') {
			errno = -5;
			return 0;
		}

		if (inbuf[1] == 'i') need = 6 - inpos;
		else if (inbuf[1] == 'd') need = 25 - inpos;
		else {
			errno = -5;
			return 0;
		}
	}

	ret = port->read((char *) inbuf + inpos, need);
	if (ret == -1) {
		errno = -5;
		return 0;
	}
	inpos += ret;

	if (inpos > 2 && ret == need) {
		if (ibutton_crc(&inbuf[1], inpos - 2) != inbuf[inpos-1]) {
			errno = -5;
			return 0;
		}
		return 1;
	}

	return 0;
}

int
iO::send(char cmd, size_t len)
{
	outbuf[0] = 'K';
	outbuf[1] = cmd;
	outbuf[len + 2] = ibutton_crc(&outbuf[1], len + 1);

	port->write((char *) outbuf, len + 3);

	return 0;
}

File: python/iO.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
#
# Copyright (C) 2008, i Bilisim Teknolojileri Ltd. Sti.
#

import time

import seriport
import crc

def CRC(data):
    c = crc.Crc(width=8,poly=0x31L,reflect_in=True,xor_in=0L,reflect_out=True,xor_out=0L)
    return "%c" % c.table_driven(data)


class iOData:
    def __init__(self, packet):
        if packet[0] != "d":
            raise ValueError, "Not a iO data packet"
        packet = map(ord, packet[1:])
        self.pin = (packet[3] << 24) + (packet[2] << 16) + (packet[1] << 8) + packet[0]
        packet = packet[4:]
        adc = zip(packet[::2], packet[1::2])
        self.adc = [(x << 8) + y for x, y in adc]


class iO:
    def __init__(self, device="/dev/ttyUSB0", speed=38400):
        self.receive(init=True)
        self.port = seriport.Port(device, speed)
        # Pull RESET to high for rev1 (FTDI design)
        self.port.rts=0
        # Give Atmel some time to setup UART
        time.sleep(0.5)
        # Continue with handshake
        self.ask_version()
        data = self.receive_block()
        if data[0] != 'i' or data[1] != 'O':
            raise IOError("Not an iO device")
        self.version = ord(data[2])
        self.revision = ord(data[3])
        if self.version != 1:
            raise IOError("Version %d of iO is not known by this driver" % self.version)
    
    def send(self, data):
        self.port.write("K%s%c" % (data, CRC(data)))
    
    def receive(self, init=False):
        if init:
            self.buffer = ""
            self.need = 2
            return
        self.buffer += self.port.read(self.need - len(self.buffer))
        size = len(self.buffer)
        if size == 2:
            if self.buffer[0] != 'R':
                raise IOError("Bogus data '%s'" % self.buffer)
            c = self.buffer[1]
            if c == 'i':
                self.need = 6
            elif c == 'd':
                self.need = 23
            else:
                raise IOError("Unknown command '%c'" % c)
        if size != self.need:
            return None
        payload = self.buffer[1:-1]
        if self.buffer[-1] != CRC(payload):
            raise IOError("CRC Error")
        self.buffer = ""
        self.need = 2
        return payload
    
    def receive_block(self):
        while True:
            data = self.receive()
            if data:
                return data
    
    def ask_version(self):
        self.send("v")
    
    def reconnect(self):
        # Send max command len of zeroes, so firmware get backs to command state
        self.port.write("\x00" * 10)
        # Flush the input buffer, it may contain previous garbage
        self.port.flush_input()
        self.receive(init=True)
        # Do the version exchange again
        self.ask_version()
        data = self.receive_block()
        if data[0] != 'i' or data[1] != 'O':
            raise IOError("Not an iO device")
        print "Reconnected"
    
    def report(self):
        self.send("r")
    
    def led(self,  a=None,  b=None):
        cmd = 0
        if not a is None:
            cmd |= 0x10
            if a:
                cmd |= 0x01
        if not b is None:
            cmd |= 0x20
            if b:
                cmd |= 0x02
        self.send("l%c" % cmd)
    
    def i2c_send(self,  addr,  data):
        pad = ""
        if len(data) < 142:
            pad = "\x00" * (142 - len(data))
        self.send("i%c%c%s" % (addr,  len(data), data + pad))
    
    def pin_in(self, pin, pullup=0):
        if pullup:
            out = 1
        else:
            out = 0
        self.send("d%c%c" % (pin, out))
    
    def pin_out(self, pin, value):
        if value:
            out = 3
        else:
            out = 2
        self.send("d%c%c" % (pin, out))
    
    def adc(self, mask):
        if mask < 0 or mask > 255:
            raise ValueError, "Invalid ADC mask"
        self.send("A%c" % mask)
    
    def pwm(self, pin, value, prescaler=None, mode=None, top=None):
        prescalers = (1, 8, 32, 64, 128, 256, 1024)
        if not pin in (1, 2, 3, 4):
            raise ValueError, "Invalid pin no"
        if value < 0 or value > 65535:
            raise ValueError, "Invalid PWM value"
        
        if prescaler or mode or top:
            # Defaults
            if not prescaler:
                prescaler = 1
            if not mode:
                mode = "fast"
            if not top:
                top = 0xffff
            # Validate
            if not prescaler in prescalers:
                raise ValueError, "Invalid prescaler"
            md = 0
            if mode == "off":
                md = 0
            elif mode == "fast":
                md = 1
            elif mode == "top":
                md = 2
            else:
                raise ValueError, "Invalid mode"
            # Set
            self.send("P%c%c%c%c%c%c%c" % (
                pin,
                prescalers.index(prescaler),
                md,
                (top&0xff00) >> 8, top&0xff,
                (value&0xff00) >> 8, value&0xff
            ))
        else:
            self.send("p%c%c%c" % (pin, (value&0xff00) >> 8, value&0xff))
